experimental/Origami/src/Origami.jl

```diff
@@ -0,0 +1,299 @@
+
+
+module OrigamiHelper
+
+using ..GAP
+
+function __init__()
+    mod_p = "https://ag-weitze-schmithusen.github.io/ModularGroup/PackageInfo.g"
+    ori_p = "https://ag-weitze-schmithusen.github.io/Origami/PackageInfo.g"
+    GAP.Packages.install(mod_p)
+    GAP.Packages.install(ori_p)
+    GAP.Packages.load("Origami")
+end
+
+end
+
+include("types.jl")
+include("canonical.jl")
+include("deck_group.jl")
+include("action.jl")
+include("special_origigami.jl")
+
+@doc raw"""
+    origami(h::PermGroupElem, v::PermGroupElem)
+
+blablabla
+# Examples
+```jldoctest
+julia> h = @perm (1,2)
+(1,2)
+
+julia> v = @perm (1,2)
+(1,2)
+
+julia> o = origami(h,v)
+Origami ((1,2),(1,2))
+```
+
+"""
+function origami(h::PermGroupElem, v::PermGroupElem)
+    deg = max(degree(h), degree(v))
+    return Origami(GAP.Globals.Origami(GapObj(h), GapObj(v)), h, v, deg)
+    # TODO check for transitivity? GAP already does this and this was so slow
+    # for huge origamis
+    #perm_group = permutation_group(deg, [h, v])
+    #if transitivity(perm_group) > 0
+        
+    #end
+end
+
+function origami_disconnected(h::PermGroupElem, v::PermGroupElem, d::Integer)
+    return Origami(GAP.Globals.OrigamiNC(GapObj(h), GapObj(v), d), h, v, d)
+end
+
+function Base.:(==)(a::Origami, b::Origami)
+    # TODO rewrite this? for now use Gap equality
+    return GAP.evalstr("$(string(a)) = $(string(b))")
+end
+
+function horizontal_perm(o::Origami)
+    return o.h
+end
+
+function vertical_perm(o::Origami)
+    return o.v
+end
+
+function degree(o::Origami)
+    return o.d
+end
+
+function Base.show(io::IO, o::Origami)
+    h = horizontal_perm(o)
+    v = vertical_perm(o)
+    d = degree(o)
+    print(io, "Origami ($(h),$(v), $(d))")
+end
+
+GapObj(O::Origami) = O.o
+
+function stratum(o::Origami)
+    # cannot use comm from Oscar because of different definitions
+    h = horizontal_perm(o)
+    v = vertical_perm(o)
+    commutator = h * v * h^-1 * v^-1
+    cycle_struc = cycle_structure(commutator)
+    different_cycle_count = length(cycle_struc)
+    stratum::Vector{Integer} = []
+    for i in 1:different_cycle_count
+        cycle_type_entry = cycle_struc[i]
+        cycle_type_length = cycle_struc[i][1]
+        if(cycle_type_entry[1] == 1)
+            # ignore cycles of length 1
+            continue
+        end
+        for j in 1:cycle_type_entry[2]
+            push!(stratum, ZZ(cycle_type_length - 1))
+        end
+    end
+    return stratum
+end
+
+function genus(o::Origami)
+    return ZZ((sum(stratum(o)) + 2) * 0.5)
+end
+
+function veech_group(O::Origami)
+    # TODO use hashtables or not? Implement modular subgroup?
+    GAP.Globals.ComputeVeechGroupWithHashTables(GapObj(O))
+end
+
+function index_monodromy_group(o::Origami)
+    GAP.Globals.IndexOfMonodromyGroup(GapObj(o))
+end
+
+function sum_of_lyapunov_exponents(o::Origami)
+    gap_obj = GAP.Globals.SumOfLyapunovExponents(GapObj(o))
+    return GAP.gap_to_julia(gap_obj)
```

Comment:
Using `gap_to_julia` is not recommended. Instead specify the desired target type. I can only guess here, but maybe the result is a (potentially big) integer, in that case:
```suggestion
    return ZZRingElem(gap_obj)
```
This has multiple advantages:
1. It makes it obvious what the result type of this function is
2. it allows Julia itself to predict the return type, and hence makes the code type stable code (which is impossible when using `gap_to_julia`)
3. it adds a type check: if something goes wrong and the GAP function returns something that isn't an integer (e.g. `fail) then we get a helpful error here instead of a confusing one elsewhere later on.